### PR TITLE
Add rate limiting in the server using built in Nest.js capability

### PR DESCRIPTION
**Pull request by jss475**
[#3545 ](https://github.com/twentyhq/twenty/issues/3545)

- Add rate limiting for graphql and http requests in the backend using built in Nest.js capability

### CODE REVIEW

File: packages/twenty-server/.env.test [hunk 1]
```diff
@@ -9,6 +9,12 @@ FRONT_BASE_URL=http://localhost:3001
 ACCESS_TOKEN_SECRET=secret_jwt
 LOGIN_TOKEN_SECRET=secret_login_tokens
 REFRESH_TOKEN_SECRET=secret_refresh_token
+LOGGED_IN_SHORT_TTL=3000
+LOGGED_IN_LONG_TTL=30000
+LOGGED_IN_SHORT_LIMIT=10
+LOGGED_IN_LONG_LIMIT=30
+NON_LOGGED_IN_TTL=3000
+NON_LOGGED_IN_LIMIT=10
```
Comment (magrinj) [inline packages/twenty-server/.env.test:17]: @jss475 Better to remove unused env variables that you add

File: packages/twenty-server/src/app.module.ts [hunk 2]
```diff
@@ -25,6 +29,15 @@ import { WorkspaceModule } from './workspace/workspace.module';
     IntegrationsModule,
     CoreModule,
     WorkspaceModule,
+    ThrottlerModule.forRootAsync({
+      useClass: ThrottlerConfigService,
+    }),
+  ],
+  providers: [
+    {
+      provide: APP_GUARD,
```
Comment (magrinj) [inline packages/twenty-server/src/app.module.ts:38]: I'm not sure if it's going to work for the flexible schema, @charlesBochet I think we need to test this

**Review by magrinj — CHANGES_REQUESTED** [review]
Thanks for your contribution !
Some changes and tests are needed before merging it 🙂

File: packages/twenty-server/src/guards/gql-throttler.guard.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { ExecutionContext, Injectable } from '@nestjs/common';
+import { GqlExecutionContext } from '@nestjs/graphql';
+import { ThrottlerGuard } from '@nestjs/throttler/dist';
```
[the remark this answers is not in the document]
Comment (jss475) [reply packages/twenty-server/src/guards/gql-throttler.guard.ts]: @magrinj I thought the same thing, but for me going into the /dist folder is the only way the imports work. Maybe it's just how my files are configured(?)

File: packages/twenty-server/src/guards/gql-throttler.guard.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+@Injectable()
+export class GqlHttpThrottlerGuard extends ThrottlerGuard {
+  getRequestResponse(context: ExecutionContext) {
```
[the remark this answers is not in the document]
Comment (jss475) [reply packages/twenty-server/src/guards/gql-throttler.guard.ts]: Good call! Thank you!

File: packages/twenty-server/src/integrations/environment/environment.service.ts [hunk 5]
```diff
@@ -281,4 +281,28 @@ export class EnvironmentService {
...
     return this.configService.get<boolean>('IS_SIGN_UP_DISABLED') ?? false;
   }
+
+  getLoggedInShortTtl(): number {
+    return this.configService.get<number>('LOGGED_IN_SHORT_TTL') ?? 1000;
+  }
+
+  getLoggedInShortLimit(): number {
+    return this.configService.get<number>('LOGGED_IN_SHORT_LIMIT') ?? 10;
+  }
```
Comment (magrinj) [inline packages/twenty-server/src/integrations/environment/environment.service.ts:291]:
```suggestion
```

File: packages/twenty-server/src/integrations/environment/environment.service.ts [hunk 6]
```diff
@@ -281,4 +281,28 @@ export class EnvironmentService {
...
+    return this.configService.get<number>('LOGGED_IN_LONG_LIMIT') ?? 20;
+  }
+
+  getNonLoggedInTtl(): number {
+    return this.configService.get<number>('NON_LOGGED_IN_TTL') ?? 1000;
+  }
+
+  getNonLoggedInLimit(): number {
+    return this.configService.get<number>('NON_LOGGED_IN_LIMIT') ?? 10;
+  }
```
Comment (magrinj) [inline packages/twenty-server/src/integrations/environment/environment.service.ts:307]:
```suggestion
```

File: packages/twenty-server/.env.example [hunk 7]
```diff
@@ -9,6 +9,7 @@ LOGIN_TOKEN_SECRET=replace_me_with_a_random_string_login
 REFRESH_TOKEN_SECRET=replace_me_with_a_random_string_refresh
 SIGN_IN_PREFILLED=true
 
+
```
Comment (magrinj) [inline packages/twenty-server/.env.example:12]:
```suggestion
# LOGGED_IN_LONG_TTL=3000
# LOGGED_IN_LONG_LIMIT=20
```

File: packages/twenty-server/.env.test [hunk 8]
```diff
@@ -11,6 +11,7 @@ LOGIN_TOKEN_SECRET=secret_login_tokens
 REFRESH_TOKEN_SECRET=secret_refresh_token
 
 
+
```
Comment (magrinj) [inline packages/twenty-server/.env.test:14]:
```suggestion
# LOGGED_IN_LONG_TTL=3000
# LOGGED_IN_LONG_LIMIT=20
```

**Review by magrinj — CHANGES_REQUESTED** [review]

**Comment by magrinj** [comment]
@jss475 Thanks for your work, I've change the code because our graphQL schema is dynamically constructed, so `@nestjs/throttler` was not handling request on the dynamic side.
Instead I've created a `graphql-yoga` plugin.
cc @charlesBochet

File: packages/twenty-server/.env.example [hunk 9]
```diff
@@ -52,3 +52,5 @@ SIGN_IN_PREFILLED=true
 # EMAIL_SMTP_USER=
 # EMAIL_SMTP_PASSWORD=
 # PASSWORD_RESET_TOKEN_EXPIRES_IN=5m
+# LOGGED_IN_LONG_TTL=
```
Comment (charlesBochet) [inline packages/twenty-server/.env.example:55]: I am not sure about this naming, reading the code to understand what it is

Comment (charlesBochet) [reply packages/twenty-server/.env.example:55]: API_RATE_LIMITING_TTL
API_RATE_LIMITING_LIMIT

**Review by charlesBochet — APPROVED** [review]
Approved with comments. We should also add the environment variables to the documentation

**Review by charlesBochet — APPROVED** [review]

**Pull request merged**